### PR TITLE
feat(udf): support inlined Rust UDF

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR allows users to inline their Rust UDF code in the create-function statement.

```sql
create function gcd(int, int) returns int language rust as $$
    fn gcd(mut a: i32, mut b: i32) -> i32 {
        while b != 0 {
            let t = b;
            b = a % b;
            a = t;
        }
        a
    }
$$;
```

The function will be [built](https://github.com/risingwavelabs/arrow-udf/blob/765aecd70223a90066dce7f9943b697d641feb96/arrow-udf-wasm/src/build.rs) into a wasm binary in the frontend. This requires the Rust toolchain (with the wasm32-wasi target) installed in frontend's environment. The build process will run for at least 10s, during which time the command is blocking from user's perspective. It may also require network access to download dependencies. Therefore, in the future we should consider how to:

1. improve the user experience while waiting for a build to complete
2. limit resources for build to prevent DoS attack

## Checklist

- [ ] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

Add support for inline Rust UDFs.

### CODE REVIEW

**Comment by BugenZhao** [comment]
This is amazing!

> This requires the Rust toolchain (with the wasm32-wasi target) installed in frontend's environment

Should we consider including it into the Docker image?

File: src/frontend/src/handler/create_function.rs [hunk 1]
```diff
@@ -173,16 +171,56 @@ pub async fn handle_create_function(
...
+            }
+            let function_body = params
+                .as_
+                .ok_or_else(|| ErrorCode::InvalidParameterValue("AS must be specified".into()))?
+                .into_string();
+            let script = format!("#[arrow_udf::function(\"{identifier}\")]\n{function_body}");
+            body = Some(function_body.clone());
+
+            let wasm_binary =
+                tokio::task::spawn_blocking(move || arrow_udf_wasm::build("", &script))
```
Comment (BugenZhao) [inline src/frontend/src/handler/create_function.rs:202]: IIUC, this means that we don't support specifying external crate dependencies now. This sounds good as the `build.rs` could run arbitrary code and is quite unsafe in a managed environment.

I'm considering whether it's possible to support a whitelist of 3rd-party crates in the future, just like how Rust's online playground does: https://arc.net/l/quote/qtopwevk

Comment (BugenZhao) [reply src/frontend/src/handler/create_function.rs:202]: Or if `wasm` can serve as a with-host-tools toolchain in the distant future... 🤣

https://doc.rust-lang.org/nightly/rustc/platform-support.html#tier-2-with-host-tools

Comment (BugenZhao) [inline src/frontend/src/handler/create_function.rs:202]: The experimental feature of cargo script would also be helpful: https://github.com/rust-lang/cargo/issues/12207

**Review by BugenZhao — APPROVED** [review]

File: src/frontend/src/handler/create_function.rs [hunk 2]
```diff
@@ -173,16 +171,56 @@ pub async fn handle_create_function(
...
+                    .context("failed to build rust function")?;
+
+            // below is the same as `wasm` language
+            let runtime = arrow_udf_wasm::Runtime::new(&wasm_binary)?;
+            check_wasm_function(&runtime, &identifier)?;
+
+            let system_params = session.env().meta_client().get_system_params().await?;
+            let object_name = format!("{:?}.wasm", md5::compute(&wasm_binary));
+            upload_wasm_binary(
+                system_params.wasm_storage_url(),
```
Comment (BugenZhao) [inline src/frontend/src/handler/create_function.rs:213]: Are we still going to deprecate the external storage for WASM binaries? I'm not sure if a Rust WASM binary fits the metadata store. 😕

File: src/frontend/src/handler/create_function.rs (hunk 1 above)
Comment (wangrunji0408) [reply src/frontend/src/handler/create_function.rs:202]: > I'm considering whether it's possible to support a whitelist of 3rd-party crates in the future, just like how Rust's online playground does: https://arc.net/l/quote/qtopwevk

Good idea! We can follow their way: pre-download all the crates that Rust playground supports and add `--offline` to `cargo build`.

Comment (wangrunji0408) [reply src/frontend/src/handler/create_function.rs:202]: Yes I have investigated it. But `cargo script` seems to only support executables, not libraries. 😕

File: src/frontend/src/handler/create_function.rs (hunk 2 above)
Comment (wangrunji0408) [reply src/frontend/src/handler/create_function.rs:213]: Yes

Comment (BugenZhao) [reply src/frontend/src/handler/create_function.rs:213]: What do you mean by "yes"? 😄

**Comment by stdrc** [comment]
> This requires the Rust toolchain (with the wasm32-wasi target) installed in frontend's environment.

Can we detect the existance of the toolchain before building? Seems in this PR the query will run for a while before reporting missing `wasm32-wasi` target.

> The build process will run for at least 10s, during which time the command is blocking from user's perspective.

Is it possible to print some notice before building or make it run in background?

**Review by stdrc — APPROVED** [review]
Looks so successful to me!

File: src/frontend/src/handler/create_function.rs (hunk 2 above)
Comment (wangrunji0408) [reply src/frontend/src/handler/create_function.rs:213]: We will soon move WASM binaries from object store to meta store. I think meta store is capable given that we will move to SQL backend soon and binaries usually won't be larger than 10MB. 😄

**Comment by wangrunji0408** [comment]
> Can we detect the existance of the toolchain before building? Seems in this PR the query will run for a while before reporting missing wasm32-wasi target.

Sure. I've added a feature to automatically install the target on build if `--offline` is not given. Besides, I also added this target to the docker image.

**Comment by wangrunji0408** [comment]
> Is it possible to print some notice before building or make it run in background?

I did a quick investigation and found [`NoticeResponse`](https://www.postgresql.org/docs/current/protocol-flow.html#PROTOCOL-ASYNC) in pgwire protocol maybe useful.

**Pull request merged**